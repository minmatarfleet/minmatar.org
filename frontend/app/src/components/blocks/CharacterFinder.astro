---
import { i18n } from '@helpers/i18n'
const { t, translatePath } = i18n(Astro.url)

import { prod_error_messages } from '@helpers/env'

import type { User } from '@dtypes/jwt'
import * as jose from 'jose'

const auth_token = Astro.cookies.has('auth_token') ? Astro.cookies.get('auth_token').value : false
const user = auth_token ? jose.decodeJwt(auth_token) as User : false

import { get_all_alliance_members } from '@helpers/fetching/corporations'
import type { CharacterBasic } from '@dtypes/layout_components'

let members:CharacterBasic[] = []
let get_groups_members_error:string | false = false
try {
    members = (auth_token && user ? await get_all_alliance_members(auth_token) : [])
    members.sort((a, b) => a.character_name.localeCompare(b.character_name))
} catch (error) {
    get_groups_members_error = prod_error_messages() ? t('get_all_corporations_members_error') : error.message
}

interface Props {
    [propName: string]: any;
}

import Flexblock from '@components/compositions/Flexblock.astro';
import FlexInline from '@components/compositions/FlexInline.astro';

import Dialog from '@components/blocks/Dialog.astro';
import Kbd from '@components/blocks/Kbd.astro';
import Input from '@components/blocks/Input.astro';
import PilotBadge from '@components/blocks/PilotBadge.astro';
import StylessButton from '@components/blocks/StylessButton.astro';

import MagnifierIcon from '@components/icons/buttons/MagnifierIcon.astro';
---

<div
    class="[ characters-finder ]"    
    x-data={`{
        search: '',
        filtered_characters: [ false ],
        characters: ${JSON.stringify(members)},
        character_filter() {
            if (this.search === '') {
                this.filtered_characters = [ false ]
                return
            }

            const filtered_characters = []

            this.characters.forEach( (character) => {
                const search = character.character_name.toLowerCase()
                
                if ( this.search !== '' && search.includes(this.search.toLowerCase()) )
                    filtered_characters.push(character.character_id)
            })
            
            this.filtered_characters = filtered_characters
        },
        async ctrl_shift_f(e) {
            if (e.ctrlKey && e.shiftKey && e.code === 'KeyF') {
                e.preventDefault()
                
                show_characters_finder().then( (character_id) => {
                    if (!character_id)
                        return

                    show_persistent_modal({
                        partial: \`${translatePath('/partials/character_modal_component/')}?character_id=\${character_id}\`,
                    })
                })
            }
        },
        init() {
            document.addEventListener('keydown', this.ctrl_shift_f, false);
        }
    }`}
>
    <Dialog
        breakout={true}
        non_centered={true}
        class="[ w-full max-w-[500px] ]"
        x-data={`{
            close_dialog() {
                if (!characters_finder_open)
                    return

                search = ''
                filtered_characters = [ false ]
                
                var clickEvent = new MouseEvent("click", {
                    "view": window,
                    "bubbles": true,
                    "cancelable": false
                });

                document.querySelector('.characters-finder-backdrop').dispatchEvent(clickEvent);
            }
        }`}
        @keyup.escape.window="close_dialog()"
        x-bind:class="(characters_finder_open ? 'open' : '')"
        x-bind:aria-hidden="characters_finder_open == false"
        thin={true}
        x-trap="characters_finder_open"
        x-bind:inert="!characters_finder_open"
    >
        <Flexblock gap="0">
            <Input
                placeholder={t('find_character_placeholder')}
                x-model="search"
                x-on:keyup="character_filter()"
                class="[ relative z-10 ]"
            >
                <MagnifierIcon slot="icon" />
            </Input>

            <div class="[ relative ]">
                <Flexblock gap='var(--space-3xs)'>
                    <Flexblock
                        class="[ character-finder-list ][ light-transparency w-full ]"
                        gap='var(--space-3xs)'
                        x-show="(search !== '')"
                    >
                        <div
                            class="[ py-[var(--space-2xs)] px-[var(--space-2xs)] ]"
                            x-show="search !== '' && filtered_characters.length === 0"
                        >
                            <span>{t('no_characters')}</span>
                        </div>
                        {members.map((member) =>
                            <StylessButton
                                type="button"
                                character-finder-action="accept"
                                character-finder-data={member.character_id}
                                x-show="show_item($el)"
                                x-on:click="(search = '')"
                                x-data={`{
                                    show_item() {
                                        return filtered_characters.includes(${member.character_id})
                                    }
                                }`}
                            >
                                <PilotBadge
                                    character_id={member.character_id}
                                    character_name={member.character_name}
                                    corporation={member.corporation}
                                />
                            </StylessButton>
                        )}
                    </Flexblock>

                    <FlexInline justification='center'>
                        <FlexInline gap="var(--space-3xs)">
                            <span class="[ text-[var(--highlight)] ]">(</span>
                            <Kbd>Shift</Kbd>
                            <span class="[ text-[var(--highlight)] ]">+</span>
                            <span class="[ text-[var(--highlight)] ]">)</span>
                            <Kbd>Tab</Kbd>
                            <small>to navigate</small>
                        </FlexInline>
                        <FlexInline gap="var(--space-3xs)">
                            <Kbd>Enter</Kbd>
                            <small>to select</small>
                        </FlexInline>
                        <FlexInline gap="var(--space-3xs)">
                            <Kbd>Esc</Kbd>
                            <small>to close</small>
                        </FlexInline>
                    </FlexInline>
                </Flexblock>
            </div>
        </Flexblock>
    </Dialog>

    <div
        class="characters-finder-backdrop"
        x-bind:class="(characters_finder_open ? 'open' : '')"
        character-finder-action="close"
        x-on:click="(filter = '')"
    />
</div>

<style lang="scss">
    :global([character-finder-action] *) {
        pointer-events: none !important;
    }

    picture {
        aspect-ratio: 1/1;
        width: 48px;
    }

    :global(.character-finder-list [style="display: none;"] ~ button) {
        margin-block-start: 0 !important;
    }

    :global(.character-finder-list :not([style="display: none;"]) ~ button) {
        margin-block-start: var(--space-3xs) !important;
    }

    .dialog {
        top: 12%;
    }
    
    .character-finder-list {
        overflow-y: auto;
        max-height: 320px;
        background-color: #171616;
        border: solid 1px var(--border-color);
        border-top: none;
        height: calc(100% - 20px);
    }

    [character-finder-action="close"] {
        position: fixed;
        inset: 0;
        z-index: var(--backdrop-z-index);
        display: none;

        &.open {
            display: block;
        }
    }

    [character-finder-action="accept"] {
        background-color: transparent;
        cursor: pointer;
        border: none;
        padding: 0;

        &:focus-within {
            outline: none;
            background-color: var(--component-background);
        }

        span {
            color: var(--highlight);
        }

        small {
            color: var(--foreground);
        }
        
        &:hover {
            background-color: var(--component-background);
        }
    }
</style>