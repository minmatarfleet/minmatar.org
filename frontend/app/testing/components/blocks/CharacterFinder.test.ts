import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test } from "vitest";
import CharacterFinder from "@components/blocks/CharacterFinder.astro";

test("CharacterFinder defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(CharacterFinder, {});

  expect(result).toMatchSnapshot();
});
