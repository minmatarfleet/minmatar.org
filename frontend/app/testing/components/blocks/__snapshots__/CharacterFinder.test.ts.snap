// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterFinder defaults 1`] = `
"<div class="[ characters-finder ]" x-data="{
        search: '',
        close_dialog() {
            characters_finder_open = false
            this.search = ''
            $refs.character_finder_list.innerHTML = ''
        },
        ctrl_shift_f(e) {
            if (e.ctrlKey &#38;&#38; e.shiftKey &#38;&#38; e.code === 'KeyF') {
                show_characters_finder()
            }
        },
        init() {
            document.addEventListener('keydown', this.ctrl_shift_f, false);
        }
    }" data-astro-cid-no7hzqam> <div role="dialog" class="dialog solid-transparency breakout non-centered fixed [ w-full max-w-[500px] ]" x-bind:class="(characters_finder_open ? 'open' : '')" x-bind:aria-hidden="characters_finder_open == false" x-trap="characters_finder_open" x-bind:inert="!characters_finder_open" @keyup.escape.window="close_dialog()" @mousedown.outside="open &#38;&#38; close_dialog()" data-astro-cid-no7hzqam="true"> <div class="wrapper" data-mod data-astro-cid-yv35cmqx style="--wrapper_padding_inline: var(--space-3xs);--wrapper_padding_inline_end: var(--space-3xs);--wrapper_padding_block: var(--space-3xs);--wrapper_max_width: 80rem;">   <div class="flex-block gap-0" data-astro-cid-no7hzqam="true" data-mod style="--flex_block_gap: 0;">  <div class="[ input-group ]" data-astro-cid-ovtlrinh style="--width: 100%;"> <svg width="24" height="24" viewBox="0 0 23 23" fill="none"> <path d="M18.5455 16.3636H17.9127L17.0291 15.48C18.0095 14.1573 18.5408 12.5555 18.5455 10.9091C18.5455 6.68727 15.1309 3.27273 10.9091 3.27273C9.27273 3.27273 7.63636 3.81818 6.31636 4.8C2.94545 7.33091 2.25818 12.12 4.78909 15.4909C7.32 18.8618 12.1091 19.5491 15.48 17.0182L16.3636 17.9018V18.5455L21.8182 24L24 21.8182L18.5455 16.3636ZM10.9091 16.3636C7.89818 16.3636 5.45455 13.92 5.45455 10.9091C5.45455 7.89818 7.89818 5.45455 10.9091 5.45455C13.92 5.45455 16.3636 7.89818 16.3636 10.9091C16.3636 13.92 13.92 16.3636 10.9091 16.3636ZM2.18182 5.45455L0 7.63636V0H7.63636L5.45455 2.18182H2.18182V5.45455ZM21.8182 0V7.63636L19.6364 5.45455V2.18182H16.3636L14.1818 0H21.8182ZM5.45455 19.6364L7.63636 21.8182H0V14.1818L2.18182 16.3636V19.6364H5.45455Z" fill="currentColor"></path> </svg> <input placeholder="Type character name to find" autocomplete="off" name="character_name" x-model="search" hx-get="/partials/character_finder_list_component/" hx-trigger="keyup changed delay:0.5s" hx-target="#character-finder-list" hx-indicator=".loader" class="[ relative z-10 ] astro-ovtlrinh" data-astro-cid-no7hzqam="true" data-astro-cid-ovtlrinh style="--width: 100%;"> </div>   <div class="[ relative ]" data-astro-cid-no7hzqam> <div class="flex-block gap---space-3xs" data-astro-cid-no7hzqam="true" data-mod style="--flex_block_gap: var(--space-3xs);">  <div class="block-list [ character-finder-list ][ light-transparency w-full ]" id="character-finder-list" x-ref="character_finder_list" data-astro-cid-no7hzqam="true" data-astro-cid-s4x3vzov style="--block_list_gap: var(--space-3xs);">  </div>   <div class="flex-inline" data-astro-cid-no7hzqam="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-s);--flex_inline_justification: center;">  <div class="flex-inline" data-astro-cid-no7hzqam="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-3xs);--flex_inline_justification: flex-start;">  <span class="[ text-[var(--highlight)] ]" data-astro-cid-no7hzqam>(</span> <kbd data-astro-cid-2fpfdw4g>Shift</kbd>  <span class="[ text-[var(--highlight)] ]" data-astro-cid-no7hzqam>+</span> <span class="[ text-[var(--highlight)] ]" data-astro-cid-no7hzqam>)</span> <kbd data-astro-cid-2fpfdw4g>Tab</kbd>  <small data-astro-cid-no7hzqam>to navigate</small>  </div>   <div class="flex-inline" data-astro-cid-no7hzqam="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-3xs);--flex_inline_justification: flex-start;">  <kbd data-astro-cid-2fpfdw4g>Enter</kbd>  <small data-astro-cid-no7hzqam>to select</small>  </div>   <div class="flex-inline" data-astro-cid-no7hzqam="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-3xs);--flex_inline_justification: flex-start;">  <kbd data-astro-cid-2fpfdw4g>Esc</kbd>  <small data-astro-cid-no7hzqam>to close</small>  </div>    </div>    </div>  </div>  </div>    </div>   </div>  </div> "
`;
